File: C++/t20.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

/**
 *  leetcode第20题
 *  利用栈结构后进先出的特性依次比对字符
 *  Author:lyc
*/



class Solution {
public:
    bool isValid(string s) {

        // 定义结果布尔变量，如果输入字符串为空直接返回true
        bool result = true;
        if(s == ""){
            return result;
        };

        // 定义括号类型map
        map<char,char> str_dic = {
            {'(',')'},
            {'[',']'},
            {'{','}'},
            {'a','a'}
        };

        // 定义栈结构
        vector<char> stack = {'a'};

        for(auto i:s){
            if(i == ' ')
            {
                // 如果该字符为空直接跳过
                continue;
            }else if(str_dic.count(stack.back()) == 1 & i == str_dic[stack.back()])
            {
                // 如果与栈中第一个元素匹配成功，则删除栈中该元素
                stack.pop_back();
            }
            else
            {
                // 如果与栈中第一个元素不匹配，则将该元素添加至栈中
                stack.push_back(i);
            }
        };

        if(stack.size() != 1){
            result = false;
        };

        return result;
        
    };
};

int main()
{
    string x = "()[]{}"; 
    Solution result;
    bool res = result.isValid(x);

    cout << res << endl;
    system("pause");
}

File: Python/t20.py
# -*- conding:utf-8 -*-
#Author:lyc
import os, sys

# 题目：
# 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串，判断字符串是否有效。
#
# 有效字符串需满足：
# 左括号必须用相同类型的右括号闭合。
# 左括号必须以正确的顺序闭合。
# 注意空字符串可被认为是有效字符串。
#
# 示例 1:
# 输入: "()"
# 输出: true
# 示例 2:
# 输入: "()[]{}"
# 输出: true
# 示例 3:
# 输入: "(]"
# 输出: false
# 示例 4:
# 输入: "([)]"
# 输出: false
# 示例 5:
# 输入: "{[]}"
# 输出: true



class Solution:
    def isValid(self, s: str) -> bool:

        if not s:
            return True

        dic = {'(':')','{':'}','[':']','aaa':'aaa'}

        # 构建栈结构,只能一端进出
        z = ['aaa']
        for i in s:
            if i == ' ':
                continue
            elif z[-1] in dic and i == dic[z[-1]]:
                z.pop()
            else:
                z.append(i)

        return True if z == ['aaa'] else False



#x = "()[]{}"
x = "([)]"
result = Solution()
print(result.isValid(x))




